src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -7,400 +7,441 @@
+      val state = get()
+      if (!state.isInstanceOf[Success[T]]) dispatchOrAddCallbacks(state, new Transformation[T, U](Xform_recover, pf, executor)) // Short-circuit if we get a Failure
+      else this.asInstanceOf[Future[U]]
+    }
+
+    override final def mapTo[S](implicit tag: scala.reflect.ClassTag[S]): Future[S] =
+      if (!get.isInstanceOf[Failure[T]]) super[Future].mapTo[S](tag) // Short-circuit if we get a Success
+      else this.asInstanceOf[Future[S]]
+
+    override final def failed: Future[Throwable] = {
+      val state = get()
+      if (!state.isInstanceOf[Success[T]]) super.failed
+        //dispatchOrAddCallbacks(state, new Transformation[T, Throwable](Xform_failed, Future.failedFun[T], InternalCallbackExecutor)) // Short-circuit if we get a Failure
+      else Future.failedFailureFuture // Cached instance in case of already known success
+    }
+
+
+    override def toString: String = toString0
+
+    @tailrec private final def toString0: String = {
+      val state = get()
+      if (state.isInstanceOf[Try[T]]) "Future("+state+")"
+      else if (state.isInstanceOf[Link[T]]) state.asInstanceOf[Link[T]].promise(this).toString0
+      else /*if (state.isInstanceOf[Callbacks[T]]) */ "Future(<not completed>)"
+    }
 
     /** Try waiting for this promise to be completed.
+     * Returns true if it completed at the end of this waiting time.
+     * Does not allow Duration.Undefined as a parameter, and throws IllegalArgumentException
+     * if it is passed in.
      */
-    protected final def tryAwait(atMost: Duration): Boolean = if (!isCompleted) {
-      import Duration.Undefined
-      atMost match {
-        case e if e eq Undefined => throw new IllegalArgumentException("cannot wait for Undefined period")
-        case Duration.Inf        =>
+    protected final def tryAwait(atMost: Duration): Boolean = tryAwait0(atMost) ne null
+
+    private[this] final def tryAwait0(atMost: Duration): Try[T] =
+      if (atMost ne Duration.Undefined) {
+        val v = value0
+        if ((v ne null) || atMost <= Duration.Zero) v
+        else {
           val l = new CompletionLatch[T]()
           onComplete(l)(InternalCallbackExecutor)
-          l.acquireSharedInterruptibly(1)
-        case Duration.MinusInf   => // Drop out
-        case f: FiniteDuration   =>
-          if (f > Duration.Zero) {
-            val l = new CompletionLatch[T]()
-            onComplete(l)(InternalCallbackExecutor)
-            l.tryAcquireSharedNanos(1, f.toNanos)
-          }
-      }
 
-      isCompleted
-    } else true // Already completed
+          if (atMost.isFinite)
+            l.tryAcquireSharedNanos(1, atMost.toNanos)
+          else
+            l.acquireSharedInterruptibly(1)
+
+          l.result
+        }
+      } else throw new IllegalArgumentException("Cannot wait for Undefined duration of time")
 
     @throws(classOf[TimeoutException])
     @throws(classOf[InterruptedException])
```

Comment:
nit: it was my impression that `@throws[TimeoutException]` was preferred now, but I could be mistaken

Reply:
@NthPortal I can definitely switch to that!